### PR TITLE
[Protocol3] (draft): implement pluggable block types

### PR DESCRIPTION
**Pull request by dong77**

### CODE REVIEW

**Comment by dong77** [comment]
@Brechtpd @letsgoustc @kongliangzhong 
I've created this draft implementation for #https://github.com/Loopring/protocols/issues/289
Please review the **overall implementation** and let me know what you think.

File: packages/loopring_v3/contracts/impl/Exchange.sol [hunk 1]
```diff
@@ -427,7 +427,41 @@ contract Exchange is IExchange, Claimable, ReentrancyGuard
...
+            let ptr := mload(0x40)
+            calldatacopy(ptr, 0, calldatasize)
+            let result := delegatecall(gas, _addr, ptr, calldatasize, 0, 0)
+            let size := returndatasize
+            returndatacopy(ptr, 0, size)
+
+            switch result
+            case 0 { revert(ptr, size) }
+            default { return(ptr, size) }
+        }
```
Comment (Brechtpd) [inline packages/loopring_v3/contracts/impl/Exchange.sol]: Delegatecall can also be done in [solidity](https://solidity.readthedocs.io/en/v0.5.10/units-and-global-variables.html?highlight=delegatecall#members-of-address-types) like this:
`<address>.delegatecall(bytes memory) returns (bool, bytes memory)`

`encodeWithSelector` can be used to encode the parameters in bytes

**Comment by Brechtpd** [comment]
I don't see any problems with the overall design.

But looking at the code I'm not sure what the pros/cons of only making the block commitment with proxys vs making the complete exchange with a proxy. A block type with an error can still risk all users funds. Hiding the comlete exchange after a proxy is more flexible, though you could say more risky (because more code can be changed) but can also be less risky because potential errors can be fixed.

File: packages/loopring_v3/contracts/impl/Exchange.sol (hunk 1 above)
Comment (dong77) [reply packages/loopring_v3/contracts/impl/Exchange.sol]: This is not important, you can optimize it later :)

**Comment by dong77** [comment]
Putting  the DEX contract behind a proxy is an easier approach but it has a drawback: when we deploy a new DEX contract to replace the old one, we have to somehow replicate the entire history, eg the virtual blockchain. This can be very expensive. Otherwise users will have to withdraw and re-deposit. 

Putting BlockProcessors behind a proxy allows us to upgrade the DEX behaviors and in the meanwhile reuse the existing virtual blockchain data. 

DEX operators can always ignore new block types if they don’t want to opt-in — all onchain block types are now allowed to be disabled.

**Comment by dong77** [comment]
A not for myself: disabling a blocktypr should still allow proof verification but not block commission.

**Comment by Brechtpd** [comment]
> Putting the DEX contract behind a proxy is an easier approach but it has a drawback: when we deploy a new DEX contract to replace the old one, we have to somehow replicate the entire history, eg the virtual blockchain. This can be very expensive. Otherwise users will have to withdraw and re-deposit.

I don't think that's true. If you [change the implementation](https://github.com/Loopring/protocols/blob/f4b9da6963cbbd861a8ff3b57dcd105839efda57/packages/loopring_v3/contracts/thirdparty/Proxy.sol#L141) all state is immediately available in the new implementation (just like it's immediately available for the block processors in this design). Only the logic changes, the state remains unchanged (though additional state can be added if I remember correctly).

**Comment by dong77** [comment]
BaseUpgradeabilityProxy is interesting. I'll try to figure out how to put it in front of a DEX instance.

**Pull request closed without merging**